File: apps/mobile/androidApp/src/main/kotlin/co/heva/piro/android/schedule/ScheduleViewModel.kt
package co.heva.piro.android.schedule

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import co.heva.piro.shared.api.PiroApiClient
import co.heva.piro.shared.model.OnCallSlot
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.LocalDate
import java.time.OffsetDateTime
import java.time.YearMonth
import java.time.ZoneId
import java.time.format.DateTimeFormatter

data class ScheduleUiState(
    val month: YearMonth = YearMonth.now(),
    val loading: Boolean = true,
    val error: String? = null,
    /** Resolved shifts intersecting the visible month. */
    val slots: List<OnCallSlot> = emptyList(),
    /** Local dates in the month that have at least one shift → the user's own color for that day. */
    val onCallDays: Map<LocalDate, String> = emptyMap(),
)

/**
 * Backs the Schedule calendar: loads the signed-in user's resolved on-call shifts for the visible month
 * and projects them onto local calendar days so the grid can mark on-call dates.
 */
class ScheduleViewModel(private val api: PiroApiClient) : ViewModel() {

    private val _state = MutableStateFlow(ScheduleUiState())
    val state: StateFlow<ScheduleUiState> = _state.asStateFlow()

    init {
        load(YearMonth.now())
    }

    fun previousMonth() = load(_state.value.month.minusMonths(1))
    fun nextMonth() = load(_state.value.month.plusMonths(1))

    private fun load(month: YearMonth) {
        viewModelScope.launch {
            _state.update { it.copy(month = month, loading = true, error = null) }
            // Pad the query by a day on each side so shifts straddling the month boundary come back.
            val from = month.atDay(1).atStartOfDay(ZoneId.systemDefault()).minusDays(1).toOffsetDateTime()
            val to = month.atEndOfMonth().atTime(23, 59, 59).atZone(ZoneId.systemDefault()).plusDays(1).toOffsetDateTime()
            try {
                val slots = api.getMyOnCallSlots(from.format(ISO), to.format(ISO))
                _state.update { it.copy(loading = false, slots = slots, onCallDays = projectDays(slots, month)) }
            } catch (e: Exception) {
                _state.update { it.copy(loading = false, error = "Could not load the schedule.") }
            }
        }
    }

    /** Marks each local date the user is on call, keyed to their avatar color for the calendar dots. */
    private fun projectDays(slots: List<OnCallSlot>, month: YearMonth): Map<LocalDate, String> {
        val days = mutableMapOf<LocalDate, String>()
        val zone = ZoneId.systemDefault()
        for (slot in slots) {
            val start = runCatching { OffsetDateTime.parse(slot.startsAt).atZoneSameInstant(zone).toLocalDate() }.getOrNull() ?: continue
            val end = runCatching { OffsetDateTime.parse(slot.endsAt).atZoneSameInstant(zone).toLocalDate() }.getOrNull() ?: continue
            var d = start
            while (!d.isAfter(end)) {
                if (YearMonth.from(d) == month) days[d] = slot.userColor.ifBlank { "#3D96FE" }
                d = d.plusDays(1)
            }
        }
        return days
    }

    private companion object {
        val ISO: DateTimeFormatter = DateTimeFormatter.ISO_OFFSET_DATE_TIME
    }
}

File: apps/mobile/iosApp/Piro/Schedule/ScheduleViewModel.swift
import Foundation
import Shared

/// Loads the signed-in user's resolved on-call shifts for the visible month and projects them onto local
/// calendar days — the SwiftUI counterpart of the Android `ScheduleViewModel`. Data comes resolved from
/// the backend (`getMyOnCallSlots`); no client-side RRULE math.
@MainActor
final class ScheduleViewModel: ObservableObject {
    @Published var monthAnchor = Date()          // any date within the visible month
    @Published var loading = true
    @Published var error: String?
    @Published var slots: [OnCallSlot] = []
    /// Local day (start-of-day) → the on-call user's avatar hex, for the calendar dots.
    @Published var onCallDays: [Date: String] = [:]

    private let api: PiroApiClient
    private let cal = Calendar.current

    init(api: PiroApiClient) {
        self.api = api
    }

    func previousMonth() {
        monthAnchor = cal.date(byAdding: .month, value: -1, to: monthAnchor) ?? monthAnchor
        Task { await load() }
    }

    func nextMonth() {
        monthAnchor = cal.date(byAdding: .month, value: 1, to: monthAnchor) ?? monthAnchor
        Task { await load() }
    }

    func load() async {
        loading = true
        error = nil
        guard let interval = cal.dateInterval(of: .month, for: monthAnchor) else { loading = false; return }
        // Pad a day on each side so shifts straddling the month boundary come back.
        let from = cal.date(byAdding: .day, value: -1, to: interval.start) ?? interval.start
        let to = cal.date(byAdding: .day, value: 1, to: interval.end) ?? interval.end
        let iso = ISO8601DateFormatter()
        do {
            let result = try await api.getMyOnCallSlots(from: iso.string(from: from), to: iso.string(from: to))
            slots = result
            onCallDays = projectDays(result)
        } catch {
            self.error = PiroError.message(error, networkFallback: "Could not load the schedule.")
        }
        loading = false
    }

    /// Marks each local date in the visible month the user is on call, keyed to their avatar color.
    private func projectDays(_ slots: [OnCallSlot]) -> [Date: String] {
        guard let month = cal.dateInterval(of: .month, for: monthAnchor) else { return [:] }
        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let isoPlain = ISO8601DateFormatter()

        var days: [Date: String] = [:]
        for slot in slots {
            let start = iso.date(from: slot.startsAt) ?? isoPlain.date(from: slot.startsAt)
            let end = iso.date(from: slot.endsAt) ?? isoPlain.date(from: slot.endsAt)
            guard let start, let end else { continue }
            var day = cal.startOfDay(for: start)
            let last = cal.startOfDay(for: end)
            while day <= last {
                if month.contains(day) { days[day] = slot.userColor.isEmpty ? "#3D96FE" : slot.userColor }
                guard let next = cal.date(byAdding: .day, value: 1, to: day) else { break }
                day = next
            }
        }
        return days
    }
}
